docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,591 @@
+- `ciphertext_size` - Size of the cipher text, or the offset into the payload
+   where *AAD* starts (as *AAD* follows the cipher text immediately).
+
+Below shows a possible `oe_seal()` implementation. Please note that GCM is
+chosen here. Other cipher modes, such as CCM and OCB, would also work. GCM wins
+the election because
+
+- It is faster than CCM.
+- It is patent free when compared to OCB.
+- It is the cipher mode chosen by the Intel SGX SDK, so is required for
+  interoperability.
+
+```C
+oe_result_t oe_seal(
+    const uint8_t* key_info,
+    size_t key_info_size,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t** blob,
+    size_t* blob_size)
+{
+    uint8_t* key = NULL;
+    size_t key_size = 0;
+
+    oe_sealed_blob_header_t* header;
+    oe_entropy_kind_t k;
+    uint8_t* payload;
+    size_t size;
+    oe_result_t result = OE_OK;
+
+    if (key_info == NULL || key_info_size != sizeof(header->key_info) ||
+        blob == NULL || blob_size == NULL)
+        return OE_INVALID_PARAMETER;
+
+    size = sizeof(*header);
+    if (size > OE_UINT32_MAX - plaintext_size)
+        return OE_INVALID_PARAMETER;
+    size += plaintext_size;
+    if (size > OE_UINT32_MAX - additional_data_size)
+        return OE_INVALID_PARAMETER;
+    size += additional_data_size;
+
+    header = (oe_sealed_blob_header_t*)oe_calloc(1, size);
+    if (header == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    oe_secure_zero_fill(header, sizeof(*header));
+
+    OE_CHECK(oe_memcpy_s(
+        &header->key_info, sizeof(header->key_info), key_info, key_info_size));
+
+    OE_CHECK(oe_get_entropy(
+        header->key_info.ki_entropy, sizeof(header->key_info.ki_entropy), &k));
+
+    OE_CHECK(oe_get_seal_key(
+        (uint8_t*)&header->key_info, key_info_size, &key, &key_size));
+
+    payload = (uint8_t*)(header + 1);
+    OE_STATIC_ASSERT(sizeof(header->tag) >= 16);
+    OE_CHECK(oe_aes_gcm_encrypt(
+        key,
+        key_size,
+        header->iv,
+        sizeof(header->iv),
+        additional_data,
+        additional_data_size,
+        plaintext,
+        plaintext_size,
+        payload,
+        header->tag));
+
+    OE_CHECK(oe_memcpy_s(
+        payload + plaintext_size,
+        additional_data_size,
+        additional_data,
+        additional_data_size));
+
+    header->ciphertext_size = (uint32_t)plaintext_size;
+    header->payload_size = (uint32_t)(plaintext_size + additional_data_size);
+
+    *blob = (uint8_t*)header;
+    *blob_size = size;
+
+done:
+    oe_free_seal_key(key, NULL);
+    if (result != OE_OK)
+        oe_free(header);
+    return result;
+}
+```
+
+The above implementation generates `key_info.ki_entropy` randomly and uses a
+constant *IV* of 0<sup>96</sup> on every invocation. The rationale behind that
+is that, per
+[NIST.SP800-38D](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf),
```

Comment:
Overall, I'm with you that we can do better with a random IV, and listed it as an [open](https://github.com/binxing/openenclave/blob/wip/sealing/docs/DesignDocs/SealingAPI.md#what-cipher-mode-to-use-in-sealing). The main reason for keeping a constant IV is for interoperability with Intel SGX SDK. 

> * Intel doesn't document it's key derivation function for seal key generation using a  NIST certified KDF. That puts a question mark on the FIPS certifiability of the implementation.

I also hope intel could have the KDF certified. However, this isn't something that can be fixed by software.

> * While the `KEYID` is a 32-byte value, it is not representative of the actual entropy with regards to the randomness of the keys. Ultimately, it is just one of many inputs (all of the others remaining constant) and the key space is only 128-bits (16 bytes). How you quantify the probability of a key derivation collision is now something that also needs to be accounted for when considering the NIST requirement.

This will **not** lower the chance of key collision **between enclaves**. But for the same enclave, the same `KEYID` always leads to the same key. Given the same *IV*, it'd be guaranteed to violate NIST requirements. Hence I randomize `KEYID` here.

> * If you take into account the birthday paradox and you're picking keys at random, you don't even get to use all 128-bits of that key space before the probability of a collision exceeds the specified threshold.

Sure. Generally speaking, there shouldn't be more than 2<sup>64</sup> 128-bit keys generated **globally**, or the chance of key collision would be considered non-negligible. The next question is how big an impact it would be should 2 keys collide. A random *IV* would mitigate the problem. And I documented that [here](https://github.com/binxing/openenclave/blob/wip/sealing/docs/DesignDocs/SealingAPI.md#what-cipher-mode-to-use-in-sealing). I personally tend to use a random *IV*, but that'd make OE blobs unable to be unsealed by Intel SGX SDK (Intel blob can still be unsealed by OE). My plan is to have a vote in the next SIG-Arch meeting.

> To be clear, it doesn't seem like a security problem since even if an attacker could get two blobs encrypted with the same key & IV and break AES-GCM, it's not particularly useful if the next blob is encrypted with a different key that you then have to go find another match for. It just makes the compliance question something we need to figure out how to actively defend.

This is a security problem indeed! Same key and IV would reveal to the adversary the GHASH key, which would allow existential forgery.

> Also, even if the protection is largely from randomly changing the key, why not just generate a random IV as well at that point? It's not like you're going to be any worse off from a security perspective.

You vote for random IV. Same do I! let's close it in our next SGX-Arch meeting.

Reply:
> This is a security problem indeed! Same key and IV would reveal to the adversary the GHASH key, which would allow existential forgery

Yes, you're absolutely right, I was only thinking about the Disclosure threat vector not the Tampering one, and so the problem is serious.